// @larscom/ngrx-store-storagesync/esm2015/lib/form-sync/form-sync.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/form-sync/form-sync.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { StoreModule } from '@ngrx/store';
import { FormGroupDirective } from './directives/form-group.directive';
import { FORM_SYNC_STORE_KEY } from './form-sync.constants';
import { FORM_SYNC_REDUCER } from './providers/form-sync.providers';
import { formSyncReducer } from './store/form.reducer';
export class FormSyncModule {
    /**
     * Import 'FormSyncModule.forRoot()' only once.
     * Use 'FormSyncModule' for additional imports
     * inside feature modules.
     * @return {?}
     */
    static forRoot() {
        return {
            ngModule: FormSyncModule,
            providers: [
                {
                    provide: FORM_SYNC_REDUCER,
                    useValue: formSyncReducer
                }
            ]
        };
    }
}
FormSyncModule.decorators = [
    { type: NgModule, args: [{
                declarations: [FormGroupDirective],
                imports: [StoreModule.forFeature(FORM_SYNC_STORE_KEY, FORM_SYNC_REDUCER)],
                exports: [FormGroupDirective]
            },] }
];
