// angular-mydatepicker/esm2015/lib/constants/constants.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Constants
 * @type {?}
 */
const D = "d";
/** @type {?} */
const DD = "dd";
/** @type {?} */
const M = "m";
/** @type {?} */
const MM = "mm";
/** @type {?} */
const MMM = "mmm";
/** @type {?} */
const Y = "y";
/** @type {?} */
const YYYY = "yyyy";
/** @type {?} */
const DATE_ROW_COUNT = 5;
/** @type {?} */
const DATE_COL_COUNT = 6;
/** @type {?} */
const MONTH_ROW_COUNT = 3;
/** @type {?} */
const MONTH_COL_COUNT = 2;
/** @type {?} */
const YEAR_ROW_COUNT = 4;
/** @type {?} */
const YEAR_COL_COUNT = 4;
/** @type {?} */
const DOT = ".";
/** @type {?} */
const UNDER_LINE = "_";
/** @type {?} */
const PIPE = "|";
/** @type {?} */
const YEAR_SEPARATOR = " - ";
/** @type {?} */
const SU = "su";
/** @type {?} */
const MO = "mo";
/** @type {?} */
const TU = "tu";
/** @type {?} */
const WE = "we";
/** @type {?} */
const TH = "th";
/** @type {?} */
const FR = "fr";
/** @type {?} */
const SA = "sa";
/** @type {?} */
const DEFAULT_LOCALE = "en";
/** @type {?} */
const ZERO_STR = "0";
/** @type {?} */
const EMPTY_STR = "";
/** @type {?} */
const SPACE_STR = " ";
/** @type {?} */
const CLICK = "click";
/** @type {?} */
const KEYUP = "keyup";
/** @type {?} */
const BLUR = "blur";
/** @type {?} */
const DISABLED = "disabled";
/** @type {?} */
const BODY = "body";
/** @type {?} */
const VALUE = "value";
/** @type {?} */
const OPTIONS = "options";
/** @type {?} */
const DEFAULT_MONTH = "defaultMonth";
/** @type {?} */
const LOCALE = "locale";
/** @type {?} */
const OBJECT = "object";
/** @type {?} */
const PX = "px";
/** @type {?} */
const STYLE = "style";
/** @type {?} */
const INNER_HTML = "innerHTML";
/** @type {?} */
const OPTS = "opts";
/** @type {?} */
const YEARS_DURATION = "yearsDuration";
/** @type {?} */
const YEARS = "years";
/** @type {?} */
const VISIBLE_MONTH = "visibleMonth";
/** @type {?} */
const SELECT_MONTH = "selectMonth";
/** @type {?} */
const SELECT_YEAR = "selectYear";
/** @type {?} */
const PREV_VIEW_DISABLED = "prevViewDisabled";
/** @type {?} */
const NEXT_VIEW_DISABLED = "nextViewDisabled";
/** @type {?} */
const DATES = "dates";
/** @type {?} */
const WEEK_DAYS = "weekDays";
/** @type {?} */
const SELECTED_DATE = "selectedDate";
/** @type {?} */
const SELECTED_DATE_RANGE = "selectedDateRange";
/** @type {?} */
const MONTHS = "months";
/** @type {?} */
const ANIMATION_END = "animationend";
/** @type {?} */
const ANIMATION_TIMEOUT = 550;
/** @type {?} */
const MY_DP_ANIMATION = "myDpAnimation";
/** @type {?} */
const ANIMATION_NAMES = ["Fade", "ScaleTop", "ScaleCenter", "Rotate", "FlipDiagonal", "Own"];
/** @type {?} */
const IN = "In";
/** @type {?} */
const OUT = "Out";
/** @type {?} */
const TABINDEX = "tabindex";
/** @type {?} */
const TD_SELECTOR = "table tbody tr td:not(.myDpDaycellWeekNbr)";
/** @type {?} */
const PREVENT_CLOSE_TIMEOUT = 50;
export { D, DD, M, MM, MMM, Y, YYYY, DATE_ROW_COUNT, DATE_COL_COUNT, MONTH_ROW_COUNT, MONTH_COL_COUNT, YEAR_ROW_COUNT, YEAR_COL_COUNT, DOT, UNDER_LINE, PIPE, YEAR_SEPARATOR, SU, MO, TU, WE, TH, FR, SA, DEFAULT_LOCALE, ZERO_STR, EMPTY_STR, SPACE_STR, CLICK, KEYUP, BLUR, DISABLED, BODY, VALUE, OPTIONS, DEFAULT_MONTH, LOCALE, OBJECT, PX, STYLE, INNER_HTML, OPTS, YEARS_DURATION, YEARS, VISIBLE_MONTH, SELECT_MONTH, SELECT_YEAR, PREV_VIEW_DISABLED, NEXT_VIEW_DISABLED, DATES, WEEK_DAYS, SELECTED_DATE, SELECTED_DATE_RANGE, MONTHS, ANIMATION_END, ANIMATION_TIMEOUT, MY_DP_ANIMATION, ANIMATION_NAMES, IN, OUT, TABINDEX, TD_SELECTOR, PREVENT_CLOSE_TIMEOUT };
