// @larscom/ngrx-store-storagesync/esm2015/lib/form-sync/store/form.reducer.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/form-sync/store/form.reducer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createReducer, on } from '@ngrx/store';
import { cloneDeep, isArray, isPlainObject, merge } from 'lodash-es';
import { deleteForm, patchForm, resetForm, setForm } from './form.actions';
/**
 * @record
 */
export function IFormSyncState() { }
// tslint:disable: no-shadowed-variable
const ɵ0 = /**
 * @param {?} state
 * @param {?} __1
 * @return {?}
 */
(state, { id, value }) => (Object.assign(Object.assign({}, state), { [id]: value })), ɵ1 = /**
 * @param {?} state
 * @param {?} __1
 * @return {?}
 */
(state, { id, value }) => merge({}, Object.assign({}, state), { [id]: value }), ɵ2 = /**
 * @param {?} state
 * @param {?} __1
 * @return {?}
 */
(state, { id }) => {
    /** @type {?} */
    const reset = (/**
     * @param {?} slice
     * @return {?}
     */
    (slice) => {
        for (const property in slice) {
            if (!slice.hasOwnProperty(property)) {
                continue;
            }
            if (isPlainObject(slice[property])) {
                reset(slice[property]);
            }
            else if (isArray(slice[property])) {
                Array(slice[property]).forEach((/**
                 * @param {?} p
                 * @return {?}
                 */
                p => reset(p)));
            }
            else {
                slice[property] = null;
            }
        }
        return slice;
    });
    /** @type {?} */
    const value = reset(cloneDeep(state[id]));
    return Object.assign(Object.assign({}, state), { [id]: value });
}, ɵ3 = /**
 * @param {?} state
 * @param {?} __1
 * @return {?}
 */
(state, { id }) => (Object.assign(Object.assign({}, state), { [id]: undefined }));
/**
 * The form sync reducer for \@ngrx/store.
 * @type {?}
 */
export const formSyncReducer = createReducer(Object(), on(setForm, (ɵ0)), on(patchForm, (ɵ1)), on(resetForm, (ɵ2)), on(deleteForm, (ɵ3)));
export { ɵ0, ɵ1, ɵ2, ɵ3 };
