// @larscom/ngrx-store-storagesync/esm2015/lib/form-sync/directives/form-group.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/form-sync/directives/form-group.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostListener, Inject, Input } from '@angular/core';
import { FormGroup } from '@angular/forms';
import { select, Store } from '@ngrx/store';
import { Subscription } from 'rxjs';
import { filter } from 'rxjs/operators';
import { FORM_SYNC_CONFIG } from '../providers/form-sync.providers';
import { patchForm } from '../store/form.actions';
import { getFormSyncValue } from '../store/form.selectors';
export class FormGroupDirective {
    /**
     * @param {?} config
     * @param {?} store
     */
    constructor(config, store) {
        this.config = config;
        this.store = store;
        this.formGroupSync = true;
        this.subscriptions = new Subscription();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (!this.formGroupId) {
            return;
        }
        const { syncOnSubmit, syncValidOnly, syncRawValue } = this.config;
        this.subscriptions.add(this.formGroup.valueChanges
            .pipe(filter((/**
         * @return {?}
         */
        () => this.formGroupSync)), filter((/**
         * @return {?}
         */
        () => !(syncValidOnly && !this.formGroup.valid))), filter((/**
         * @return {?}
         */
        () => !syncOnSubmit)))
            .subscribe((/**
         * @return {?}
         */
        () => this.dispatch(syncRawValue))));
        this.subscriptions.add(this.store
            .pipe(filter((/**
         * @return {?}
         */
        () => this.formGroupSync)), select(getFormSyncValue, { id: this.formGroupId }), filter((/**
         * @param {?} value
         * @return {?}
         */
        value => value != null)))
            .subscribe((/**
         * @param {?} value
         * @return {?}
         */
        value => this.formGroup.patchValue(value, { emitEvent: false }))));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.subscriptions.unsubscribe();
    }
    /**
     * @return {?}
     */
    onSubmit() {
        if (!this.formGroupId || !this.formGroupSync) {
            return;
        }
        const { syncOnSubmit, syncValidOnly, syncRawValue } = this.config;
        if (!syncOnSubmit) {
            return;
        }
        if (syncValidOnly && !this.formGroup.valid) {
            return;
        }
        this.dispatch(syncRawValue);
    }
    /**
     * @private
     * @param {?} syncRawValue
     * @return {?}
     */
    dispatch(syncRawValue) {
        /** @type {?} */
        const value = syncRawValue ? this.formGroup.getRawValue() : this.formGroup.value;
        this.store.dispatch(patchForm({ id: this.formGroupId, value }));
    }
}
FormGroupDirective.decorators = [
    { type: Directive, args: [{
                selector: '[formGroup]'
            },] }
];
/** @nocollapse */
FormGroupDirective.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [FORM_SYNC_CONFIG,] }] },
    { type: Store }
];
FormGroupDirective.propDecorators = {
    formGroup: [{ type: Input }],
    formGroupId: [{ type: Input }],
    formGroupSync: [{ type: Input }],
    onSubmit: [{ type: HostListener, args: ['submit',] }]
};
if (false) {
    /** @type {?} */
    FormGroupDirective.prototype.formGroup;
    /** @type {?} */
    FormGroupDirective.prototype.formGroupId;
    /** @type {?} */
    FormGroupDirective.prototype.formGroupSync;
    /**
     * @type {?}
     * @private
     */
    FormGroupDirective.prototype.subscriptions;
    /**
     * @type {?}
     * @private
     */
    FormGroupDirective.prototype.config;
    /**
     * @type {?}
     * @private
     */
    FormGroupDirective.prototype.store;
}
